### PR TITLE
Homework/sriks1981 analytical patterns

### PR DESCRIPTION
**Pull request by sriks1981**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You've done a good job using Common Table Expressions (CTEs) to separate the logic for the previous and current seasons, which makes your code easier to read and understand. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the prompt:** Your solution does not include the `first_active_season`, `last_active_season`, and `seasons_active` fields that are present in the example solution. These fields are important for tracking a player's career over time, and they should be included in your final table. You can calculate these fields similarly to how it's done in the example solution.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding more comments to explain your thought process, especially in the `CASE` statement where you determine the `change_tracking` field. This would make it easier for others to understand your code.

3) **Robustness and error handling:** Your code does not handle the case where a player's `current_season` is NULL in both the `last_season` and `current_season` tables. This could potentially lead to errors or incorrect results. You should add a condition to handle this case.

4) **Modularity and reusability:** Your code is fairly modular, with separate CTEs for the previous and current seasons. However, you could improve the reusability of your code by creating a function or stored procedure that takes the season year as a parameter and returns the state change tracking for that year. This would allow you to easily reuse your code for different seasons.

Here's an example of how you could modify your code to address these points:

```sql
-- Build a CTE of previous season
WITH last_season AS (
    SELECT * 
    FROM bootcamp.nba_players
    WHERE current_season = 1999
),
-- CTE for current season
current_season AS (
    SELECT * 
    FROM bootcamp.nba_players
    WHERE current_season = 2000
),
-- CTE for final result
result AS (
    SELECT
        coalesce(l.player_name, c.player_name) AS player_name,
        coalesce(l.first_active_season, c.current_season) AS first_active_season,
        coalesce(c.current_season, l.last_active_season) AS last_active_season,
        CASE
            WHEN l.current_season IS NULL and c.current_season IS NOT NULL and l.seasons IS NULL THEN 'New'
            WHEN l.current_season IS NOT NULL and c.current_season IS NULL THEN 'Retired'
            WHEN l.current_season IS NOT NULL and c.current_season IS NOT NULL THEN 'Continued Playing'
            WHEN l.current_season IS NULL and c.current_season IS NOT NULL and l.seasons IS NOT NULL THEN 'Returned FROM Retirement'
            ELSE 'Stayed Retired'
        END AS change_tracking,
        coalesce(l.seasons, array[]::integer[]) || 
            CASE WHEN c.current_season IS NOT NULL THEN array[c.current_season] ELSE array[]::integer[] END AS seasons_active,
        c.current_season
    FROM last_season l
    FULL OUTER JOIN current_season c
        ON c.player_name = l.player_name
)
SELECT 
    *
FROM result
```

Keep up the good work, and keep these points in mind for your future assignments!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used CTEs (Common Table Expressions) to separate the logic for the previous and current seasons, which makes your code easier to read and understand. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the homework prompt:** Your solution does not cover all aspects of the homework prompt. The prompt asks for a state change-tracking field that takes on the following values: `New`, `Retired`, `Continued Playing`, `Returned from Retirement`, and `Stayed Retired`. However, your solution does not correctly handle the `Returned from Retirement` and `Stayed Retired` states. 

In your code, you have considered a player as `Returned from Retirement` if they played after a gap of less than 3 years. However, the prompt does not specify a gap of 3 years for a player to be considered as `Returned from Retirement`. Instead, a player should be considered as `Returned from Retirement` if they were not active in the previous season but are active in the current season. 

Similarly, for the `Stayed Retired` state, your code considers all players who do not fall into the other categories as `Stayed Retired`. However, according to the prompt, a player should be considered as `Stayed Retired` if they were not active in the previous season and are still not active in the current season.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is good practice. However, you could improve the readability of your code by adding more comments to explain the logic behind your CASE statements.

3) **Robustness and error handling:** Your code does not handle potential errors or edge cases. For example, what happens if the `current_season` or `player_name` fields are NULL? You should include error handling to account for these scenarios.

4) **Modularity and reusability:** Your code is modular and the use of CTEs makes it easy to understand and reuse specific components in other contexts. However, you could improve the reusability of your code by creating a function or stored procedure that takes the season as a parameter and returns the state change-tracking field for that season.

Here's a suggestion on how you could modify your code to address these issues:

```sql
-- CTE for final result
result AS (
    select 
        coalesce(y.player_name, t.player_name) AS player_name,
        coalesce(y.first_active_season, t.current_season) AS first_active_season,
        coalesce(t.current_season, y.last_active_season) AS last_active_season,
        CASE
            -- New player - No record for last season
            WHEN y.current_season IS NULL THEN ARRAY[t.current_season]
            -- Retired player - No record for season
            WHEN t.current_season IS NULL THEN y.seasons
            -- Default - Continuing player
            else y.seasons || ARRAY[t.current_season]
        END AS seasons_active,
        CASE
            -- New Players - First active season is empty and current season is not
            WHEN y.first_active_season IS NULL and t.current_season IS NOT NULL THEN 'New'
            -- Retired players - Whose last active season is the same as the last season
            WHEN t.current_season IS NULL AND y.last_active_season = y.current_season THEN 'Retired'
            -- Active players - Ones who played in last season and are playing in current one as well
            WHEN t.current_season - y.last_active_season = 1 THEN 'Continued Playing'
            -- Returning players - Ones who played after a gap of more than 1 year
            WHEN y.last_active_season < t.current_season - 1 THEN 'Returned from Retirement'
            -- All retired players
            WHEN t.current_season IS NULL AND y.last_active_season < y.current_season THEN 'Stayed Retired'
        END AS change_tracking 
    FROM last_season y FULL OUTER JOIN current_season t
        ON t.player_name = y.player_name
)
SELECT 
    *
FROM result
```

Keep up the good work and continue practicing your SQL skills!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used Common Table Expressions (CTEs) to separate the logic for the previous and current seasons, which makes your code easier to read and understand. However, there are a few areas where your solution could be improved:

1) **Covering all aspects of the prompt:** Your solution does not cover all aspects of the homework prompt. The prompt asks for a state change-tracking field that takes on five different values, but your solution only covers four of these. You have not included a condition for players who stay out of the league, which should be labeled as 'Stayed Retired'. 

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding more comments to explain your thought process, especially in the complex case statements. 

3) **Robustness and error handling:** Your code does not handle the case where a player stays out of the league ('Stayed Retired'). This could lead to incorrect results if a player does not play for more than one season. 

4) **Modularity and reusability:** Your use of CTEs makes your code more modular and easier to understand. However, your code is not very reusable because it is specific to the 2000 and 2001 seasons. You could make your code more reusable by parameterizing the season years.

Here's how you could modify your code to address these issues:

```sql
-- CTE for final result
result AS (
    select 
        coalesce(y.player_name, t.player_name) AS player_name,
        coalesce(y.first_active_season, t.current_season) AS first_active_season,
        coalesce(t.current_season, y.last_active_season) AS last_active_season,
        CASE
            -- New player - No record for last season
            WHEN y.current_season IS NULL THEN ARRAY[t.current_season]
            -- Retired player - No record for season
            WHEN t.current_season IS NULL THEN y.seasons
            -- Default - Continuing player
            else y.seasons || ARRAY[t.current_season]
        END AS seasons_active,
        CASE
            -- New Players - First active season is empty and current season is not
            WHEN y.first_active_season IS NULL and t.current_season IS NOT NULL THEN 'New'
            -- Retired players - Who did not play current season but played in the last
            WHEN t.current_season IS NULL and y.last_active_season = y.current_season THEN 'Retired'
            -- Active players - Ones who played in last season and are playing in current one as well
            WHEN t.current_season - y.last_active_season = 1 THEN 'Continued Playing'
            -- Returning players - Ones who were inactive last season and active the current season
            WHEN y.last_active_season < t.current_season - 1 THEN 'Returned from Retirement'
            -- All players who stayed retired
            WHEN t.current_season IS NULL AND y.last_active_season < y.current_season THEN 'Stayed Retired'
            -- Add condition for players who stay out of the league
            WHEN t.current_season IS NULL AND y.last_active_season = y.current_season THEN 'Stayed Retired'
        END AS change_tracking 
    FROM last_season y FULL OUTER JOIN current_season t
        ON t.player_name = y.player_name
)
SELECT 
    *
FROM result
```

Keep up the good work, and remember to thoroughly read the prompt to ensure you're covering all required aspects.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to perform aggregations along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Missing Aspects of the Prompt:** The homework prompt asks for the aggregation of total player points and total team wins. While you have correctly aggregated total player points, you have not included the total team wins in your solution. Instead, you have calculated the total games, which was not required by the prompt. Here's how you could modify your code to include total team wins:

```sql
SUM(CASE 
    WHEN visitor_team_id = gd.team_id AND home_team_wins = 1 THEN 0
    WHEN visitor_team_id = gd.team_id AND home_team_wins = 0 THEN 1
    ELSE g.home_team_wins
END) AS total_team_wins
```

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process. However, you could improve the readability of your code by adding a few more comments to explain the logic behind your calculations, especially for the `GROUPING SETS` clause.

3) **Robustness and Error Handling:** Your code does not include any error handling or checks for invalid inputs. It would be beneficial to add some checks to ensure that the data being processed is valid. For example, you could add a check to ensure that the `game_id`, `player_name`, and `team_abbreviation` are not null before performing the aggregations.

4) **Modularity and Reusability:** Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of the `GROUPING SETS` clause in your query makes it easy to modify the dimensions along which the data is aggregated.

In conclusion, your solution is a good start, but there are a few areas where it could be improved to fully meet the requirements of the homework prompt. Keep up the good work!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used Common Table Expressions (CTEs) to separate the logic for the previous and current seasons, which makes your code easier to read and understand. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the prompt:** Your solution does not include a `season` field in the final result set. The prompt asks for a state change-tracking field for each season, so you should include a `season` field in your final result set. You can use the `current_season` field from the `current_season` CTE for this.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding more comments to explain your thought process, especially in the `CASE` statements where you determine the `seasons_active` and `change_tracking` fields.

3) **Robustness and error handling:** Your code does not handle the case where a player's `current_season` is NULL in both the `last_season` and `current_season` CTEs. This could potentially lead to NULL values in your final result set. You should add error handling to account for this case.

4) **Modularity and reusability:** Your code is modular and the use of CTEs makes it easy to understand and reuse specific components. However, you could improve the reusability of your code by parameterizing the seasons in your CTEs. This would allow you to easily change the seasons you are comparing without having to modify the main body of your code.

Here's how you could modify your code to address these points:

```sql
-- Build a CTE of previous season
WITH last_season AS (
    SELECT
        *
    FROM srik1981.bootcamp_nba_players
    WHERE current_season = 2000
),
-- CTE for current season
current_season AS (
    SELECT
        n.player_name,
        n.current_season
    FROM bootcamp.nba_players n
    WHERE current_season = 2001
),
-- CTE for final result
result AS (
    select 
        coalesce(y.player_name, t.player_name) AS player_name,
        coalesce(y.first_active_season, t.current_season) AS first_active_season,
        coalesce(t.current_season, y.last_active_season) AS last_active_season,
        CASE
            WHEN y.current_season IS NULL THEN ARRAY[t.current_season]
            WHEN t.current_season IS NULL THEN y.seasons
            else y.seasons || ARRAY[t.current_season]
        END AS seasons_active,
        CASE
            WHEN y.first_active_season IS NULL and t.current_season IS NOT NULL THEN 'New'
            WHEN t.current_season IS NULL and y.last_active_season = y.current_season THEN 'Retired'
            WHEN t.current_season - y.last_active_season = 1 THEN 'Continued Playing'
            WHEN y.last_active_season < t.current_season - 1 THEN 'Returned from Retirement'
            WHEN t.current_season IS NULL AND y.last_active_season <= y.current_season THEN 'Stayed Retired'
        END AS change_tracking,
        coalesce(t.current_season, y.current_season + 1) AS season
    FROM last_season y FULL OUTER JOIN current_season t
        ON t.player_name = y.player_name
)
SELECT 
    *
FROM result
```

Keep up the good work! Your understanding of SQL and ability to write complex queries is clear. With a few minor adjustments, your solution will fully meet the requirements of the prompt.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You've correctly used the `GROUPING SETS` function to aggregate the data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the prompt:** Your solution does not include the creation of a new table `game_details_dashboard` as specified in the prompt. You can achieve this by adding `CREATE TABLE game_details_dashboard AS` at the beginning of your query.

2) **Readability:** Your code is generally well-formatted and easy to read. However, it would be beneficial to include more comments throughout your code to explain your thought process, especially when you're defining the `data` CTE and calculating `total_team_wins`.

3) **Robustness and error handling:** Your solution does not handle the potential duplication of records. In the example solution, the `ROW_NUMBER()` function is used to deduplicate records based on `game_id`, `team_id`, and `player_id`. Consider incorporating a similar approach in your solution.

4) **Modularity and reusability:** Your code is modular and the use of a CTE at the beginning makes it easy to understand and modify if needed. However, the `GROUPING SETS` part of your query could be made more reusable by defining the grouping sets as a separate variable or CTE.

Here's a revised version of your solution incorporating these suggestions:

```sql
-- Create the new table as specified in the prompt
CREATE TABLE game_details_dashboard AS
-- Build the game data as CTE
WITH data AS (
    SELECT
        COALESCE(gd.player_name, 'N/A') AS player_name,
        coalesce(gd.team_abbreviation, 'N/A') AS team,
        coalesce(g.season, 0) AS season,
        gd.pts,
        g.visitor_team_id,
        gd.team_id,
        g.home_team_wins,
        g.game_id,
        -- Deduplicate records based on game_id, team_id, and player_id
        ROW_NUMBER() OVER(PARTITION BY gd.game_id, gd.team_id, gd.player_id ORDER BY g.game_date_est) AS row_num
    FROM bootcamp.nba_game_details gd
    JOIN bootcamp.nba_games g ON g.game_id = gd.game_id  
)
SELECT
    -- The Player Name
    coalesce(player_name, '(overall)') AS player_name,
    -- The Team Name 
    coalesce(team, '(overall)') AS team,
    -- Season
    coalesce(season, -1) AS season,
    -- Total Points
    SUM(pts) AS total_pts,
    -- Total wins
    SUM(
        CASE 
            WHEN visitor_team_id = team_id AND home_team_wins = 1 THEN 0
            WHEN visitor_team_id = team_id AND home_team_wins = 0 THEN 1
            ELSE home_team_wins
        END
    ) AS total_team_wins
FROM data
WHERE row_num = 1
GROUP BY GROUPING SETS (
    -- Grouping sets to get the aggregates
    (player_name, team),
    (player_name, season),
    (team)
)
```

Keep up the good work! With a few adjustments, your solution will be even better.

### Feedback for `query_3.sql`

Overall, your code is well-structured and logically follows the homework prompt. You've done a good job of using a Common Table Expression (CTE) to prepare the data and then applying `GROUPING SETS` to get the desired aggregates. However, there are a few areas where improvements can be made.

1) **Coverage of the homework prompt:** The prompt asks for the player who scored the most points playing for a single team. Your code does provide this, but it also includes the season in the grouping sets, which is not required for this specific question. You can simplify your code by removing the `(player_name, season)` grouping set.

2) **Readability:** Your code is generally well-commented, which makes it easier to understand your thought process. However, the comments could be more concise and directly related to the code they are explaining. For example, the comment `-- The Player Name` is not necessary as the code `coalesce(player_name, '(overall)') AS player_name` is self-explanatory. 

3) **Robustness and error handling:** Your code handles null values well by using the `COALESCE` function. However, it's not clear why you're coalescing the `season` to 0 in the CTE. If the season is not relevant to this query, it might be better to remove it altogether to avoid confusion.

4) **Modularity and reusability:** Your use of CTEs and a separate aggregation step makes your code modular and potentially reusable. However, the `aggr` CTE includes calculations for total team wins, which are not used in the final query. Removing unnecessary calculations would make your code more efficient and focused on the task at hand.

Here's a simplified version of your code:

```sql
WITH data AS (
    SELECT
        COALESCE(gd.player_name, 'N/A') AS player_name,
        coalesce(gd.team_abbreviation, 'N/A') AS team,
        gd.pts
    FROM bootcamp.nba_game_details gd
    JOIN bootcamp.nba_games g ON g.game_id = gd.game_id  
),
aggr as (
    SELECT
        coalesce(player_name, '(overall)') AS player_name,
        coalesce(team, '(overall)') AS team,
        SUM(pts) AS total_pts
    FROM data
    GROUP BY GROUPING SETS (
        (player_name, team),
        (team)
    )
)
SELECT 
    player_name, 
    team, 
    total_pts
FROM aggr
WHERE player_name <> '(overall)' 
    AND team <> '(overall)' 
ORDER BY total_pts DESC 
LIMIT 1 
```

Keep up the good work, and continue to focus on writing clean, efficient, and readable code.

### Feedback for `query_4.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You've successfully written a query to answer the question: "Which player scored the most points in one season?" However, there are a few areas where your code could be improved for better readability, robustness, and reusability.

1) **Readability:** Your code is generally well-commented, which is great. However, the comments could be more descriptive to provide a better understanding of your thought process. For example, instead of commenting `-- Build the game data as CTE`, you could explain why you're building a CTE and what it contains. Similarly, for the `aggr` CTE, you could explain that you're aggregating the data by different grouping sets to calculate total points and total team wins.

2) **Robustness and Error Handling:** Your code seems to handle edge cases well, such as when `player_name` or `team` is null. However, it's not clear how your code would handle potential errors. For example, what if `pts` is null or negative? It would be good to add some error handling or checks for these cases.

3) **Modularity and Reusability:** Your code is modular and reusable, which is excellent. The use of CTEs (`data` and `aggr`) makes your code more readable and allows for easy modification if needed. However, you could improve the reusability of your code by creating a function or stored procedure that takes in parameters (e.g., season, player name) and returns the desired result. This would allow you to reuse the same logic for different players or seasons without having to rewrite the entire query.

4) **Code Coverage:** Your code covers all aspects of the homework prompt. However, it's not clear why you're calculating `total_team_wins` in the `aggr` CTE, as this doesn't seem to be required by the prompt. If this is necessary for your solution, please explain why in your comments. If not, you could remove this to simplify your code.

In conclusion, your solution is on the right track. With a few improvements, your code could be more readable, robust, and reusable. Keep up the good work!

### Feedback for `query_5.sql`

Your solution is well-structured and logically follows the given homework prompt. You have successfully written a query to answer the question: "Which team has won the most games". Your use of Common Table Expressions (CTEs) and `GROUPING SETS` is commendable as it makes your code more readable and modular. 

Here are a few points of feedback:

1) **Readability and Comments:** Your code is generally well-commented, which is great. However, some comments could be more descriptive. For example, the comment `-- Build the game data as CTE` could be expanded to explain why you're building a CTE and what it contains. Similarly, the comment `-- Grouping sets to get the aggregates` could be more descriptive about what aggregates you're getting and why.

2) **Variable Names:** Your variable names are mostly clear and descriptive. However, the variable `aggr` could be more descriptive. Consider renaming it to something like `team_aggregates` to better reflect its contents.

3) **Robustness and Error Handling:** Your code seems to handle edge cases well, especially with the use of `COALESCE` to handle potential NULL values. However, it's not clear how your code would handle invalid inputs or potential errors. Consider adding some error handling or input validation to make your code more robust.

4) **Modularity and Reusability:** Your use of CTEs and `GROUPING SETS` makes your code modular and potentially reusable. However, consider encapsulating your code in a function or stored procedure if you find yourself writing similar queries often. This would make your code even more reusable and easier to maintain.

5) **Code Structure:** Your code structure is generally good, but there's room for improvement. For example, the `CASE` statement in the `SUM` function is a bit hard to read. Consider breaking it out into its own CTE or subquery to improve readability.

Overall, great job on this assignment! With a few minor improvements, your code will be even better. Keep up the good work!

**Review by thyanhbui1412 — APPROVED** [review]
Good job!
